guides/customizing_saving_and_serialization.py

```diff
@@ -0,0 +1,328 @@
+`model.save()`. You can override them to take full control of the state saving process.
+
+"""
+
+"""
+
+
+### `save_own_variables()` and `load_own_variables()`
+
+These methods save and load the state variables of the layer when `model.save()` and
+`keras.models.load_model()` are called, respectively. By default, the state variables
+saved and loaded are the weights of the layer (both trainable and non-trainable). Here is
+the default implementation of `save_own_variables()`:
+
+```python
+def save_own_variables(self, store):
+    all_vars = self._trainable_weights + self._non_trainable_weights
+    for i, v in enumerate(all_vars):
+        store[f"{i}"] = v.numpy()
+```
+
+The store used by these methods is a dictionary that can be populated with the layer
+variables. Let's take a look at an example customizing this.
+
+**Example:**
+"""
+
+
+@keras.utils.register_keras_serializable(package="my_custom_package")
+class LayerWithCustomVariables(keras.layers.Dense):
+    def __init__(self, units, **kwargs):
+        super().__init__(units, **kwargs)
+        self.stored_variables = tf.Variable(np.random.random((10,)), name="special_arr", dtype=tf.float32)
+
+    def save_own_variables(self, store):
+        # Stores the value of the `tf.Variable` upon saving
+        store["variables"] = self.stored_variables.numpy()
+
+    def load_own_variables(self, store):
+        # Assigns the value of the `tf.Variable` upon loading
+        self.stored_variables.assign(store["variables"])
+
+    def call(self, inputs):
+        return super().call(inputs) * self.stored_variables
+
+
+@keras.utils.register_keras_serializable(package="my_custom_package")
+class ModelWithCustomVariables(keras.Model):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.custom_dense = LayerWithCustomVariables(1)
+
+    def call(self, inputs):
+        return self.custom_dense(inputs)
+
+
+model = ModelWithCustomVariables()
+
+ref_input = np.random.random((8, 10))
+ref_output = np.random.random((8,))
+model.compile(optimizer="adam", loss="mean_squared_error")
+model.fit(ref_input, ref_output)
+
+model.save("custom_vars_model.keras")
+restored_model = keras.models.load_model("custom_vars_model.keras")
+
+np.testing.assert_allclose(
+    model.custom_dense.stored_variables.numpy(),
+    restored_model.custom_dense.stored_variables.numpy(),
+)
+
+"""
+### `save_assets()` and `load_assets()`
+
+These methods can be added to your model class definition to store and load any
+additional information that your model needs.
+
+For example, NLP domain layers such as TextVectorization layers and IndexLookup layers
+may need to store their associated vocabulary (or lookup table) in a text file upon
+saving.
+
+Let's take at the basics of this workflow with a simple file `assets.txt`.
+
+**Example:**
+"""
+
+
+@keras.saving.register_keras_serializable(package="my_custom_package")
```

Comment:
Setting the asset in build, then replacing it in load seems less clear than this could be. 

Would it be clearer to just start by setting the assets in init?:

```
class LayerWithCustomAssets(keras.layers.Dense):
  def __init__(self, vocab=None):
    self.vocab = vocab
```

That would make it clearer that you're not using the hard-coded value.

Reply:
I've updated it to use the `__init__` instead and removed the `build` override.